File: me_daily/2025/10/1030/cfgym105833/cfgym105833.cpp
#include <bits/stdc++.h>
using namespace std;

#define ALL(v) (v).begin(), (v).end()
#define For(i, n) for (int i = 0; i < (n); ++i)
#define FOR(i, a, b) for (int i = (a); i < (b); ++i)
#define pb push_back
#define fi first
#define se second
#define V vector
#define SORT(v) sort(ALL(v))

using umpii = unordered_map<int, int>;
using ll = long long;
using pii = pair<int,int>;
using vi = vector<int>;
using vll = vector<ll>;

const int inf = 0x3f3f3f3f;
const ll infl = 0x3f3f3f3f3f3f3f3fll;
const int mod = 1e9 + 7;

int fastio_init = [](){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 0;
}();

int main() {
    int n, m, k;
    cin >> n >> m >> k;
    V<V<pii>> adj(n + 1);
    V<array<int, 3>> edges;
    For(i, m) {
        int x, y, c; cin >> x >> y >> c;
        adj[x].pb({y, c});
        adj[y].pb({x, c});
        edges.pb({x, y, c});
    }
    // Dijkstra: SP_sum from 1 to n
    vll dist(n + 1, infl);
    dist[1] = 0;
    // min-heap: greater<pii> (dist, node)
    priority_queue<pii, V<pii>, greater<pii>> pq;
    pq.emplace(0, 1);
    while (!pq.empty()) {
        auto [d, u] = pq.top(); pq.pop();
        if (d > dist[u]) continue;  // 旧节点跳过
        for (auto& [v, c] : adj[u]) {
            if (dist[v] > dist[u] + c) {
                dist[v] = dist[u] + c;
                pq.emplace(dist[v], v);
            }
        }
    }
    ll sp_sum = dist[n];  // 策略1: SP总和

    vi vis(n + 1, false);
    vis[1] = true; 
    deque<int> q;
    q.emplace_back(1);
    while (!q.empty()) {
        auto u = q.front(); q.pop_front();
        for (auto& [v, _] : adj[u]) {
            if (!vis[v]) {
                vis[v] = true;
                q.emplace_back(v);
            }
        }
    }

    vll cost;
    for (auto& [x, y, c] : edges) {
        if (vis[x] && vis[y]) {
            cost.pb(c);
        }
    }
    SORT(cost);
    size_t t = min(cost.size(), (size_t)k); 
    ll strat2 = accumulate(cost.begin(), cost.begin() + t, 0LL);  // 0LL init

    // ans = min(SP, K小sum)
    cout << min(sp_sum, strat2) << '\n'; 
    return 0;
}

File: me_daily/2025/10/1030/cfgym105833/cfgym105833.py
from collections import defaultdict
from typing import List
import sys
import heapq
from collections import deque
from collections import Counter
input = lambda: sys.stdin.readline().strip()
II = lambda: int(input())
LII = lambda: list(map(int, input().split()))
SI = lambda: input()
LSI = lambda: input().split()




n, m, k = LII()
adj = [[] for _ in range(n + 1)]
edges = []
for _ in range(m):
    x, y, c = LII()
    edges.append((x, y, c))
    adj[x].append((y, c))
    adj[y].append((x, c))

inf = 10 ** 18

dis = [inf] * (n + 1)
dis[1] = 0

pq = [(0, 1)]
heapq.heapify(pq)

while pq:
    d, u = heapq.heappop(pq)
    if d > dis[u]: continue
    for v, c in adj[u]:
        if dis[v] > dis[u] + c:
            dis[v] = dis[u] + c
            heapq.heappush(pq, (dis[v], v))

sp_sum = dis[n]
vis = [False] * (n + 1)
q = deque([1])

while q:
    u = q.popleft()
    for v, _ in adj[u]:
        if not vis[v]:
            vis[v] = True
            q.append(v)

cost = []
for x, y, c in edges:
    if vis[x] and vis[y]:
        cost.append(c)

cost.sort()
t = min(k, len(cost))

start2 = sum(cost[:t])

print(min(sp_sum, start2))



 